File: 0752-open-the-lock/0752-open-the-lock.py
# BFS
class Solution:
  def openLock(self, deadends: List[str], target: str) -> int:
    target = tuple([int(d) for d in target])

    visited = set()
    for deadend in deadends:
      visited.add(tuple([int(d) for d in deadend]))

    queue = deque()
    if (0,0,0,0) not in visited:
      queue.append(((0,0,0,0), 0))
      visited.add((0,0,0,0))

    while queue:
      point, step = queue.popleft()
      if point == target:
        return step
      step += 1
      points = [
        (abs((point[0] + 1) % 10), point[1], point[2], point[3]),
        (abs((point[0] - 1) % 10), point[1], point[2], point[3]),
        (point[0], abs((point[1] + 1) % 10), point[2], point[3]),
        (point[0], abs((point[1] - 1) % 10), point[2], point[3]),
        (point[0], point[1], abs((point[2] + 1) % 10), point[3]),
        (point[0], point[1], abs((point[2] - 1) % 10), point[3]),
        (point[0], point[1], point[2], abs((point[3] + 1) % 10)),
        (point[0], point[1], point[2], abs((point[3] - 1) % 10))
      ]
      for _point in points:
        if _point not in visited:
          visited.add(_point)
          queue.append((_point, step))

    return -1

# A*
# class Solution:
#   def openLock(self, deadends: List[str], target: str) -> int:
#     target = tuple([int(d) for d in target])

#     visited = set()
#     for deadend in deadends:
#       visited.add(tuple([int(d) for d in deadend]))

#     queue = []
#     if (0,0,0,0) not in visited:
#       heapq.heappush(queue, (0, (0,0,0,0), 0))
#       visited.add((0,0,0,0))

#     while queue:
#       _, point, step = heapq.heappop(queue)
#       if point == target:
#         return step
#       step += 1
#       points = [
#         (abs((point[0] + 1) % 10), point[1], point[2], point[3]),
#         (abs((point[0] - 1) % 10), point[1], point[2], point[3]),
#         (point[0], abs((point[1] + 1) % 10), point[2], point[3]),
#         (point[0], abs((point[1] - 1) % 10), point[2], point[3]),
#         (point[0], point[1], abs((point[2] + 1) % 10), point[3]),
#         (point[0], point[1], abs((point[2] - 1) % 10), point[3]),
#         (point[0], point[1], point[2], abs((point[3] + 1) % 10)),
#         (point[0], point[1], point[2], abs((point[3] - 1) % 10))
#       ]
#       for _point in points:
#         if _point not in visited:
#           cost = step
#           for i in range(4):
#             cost += min(abs(_point[i] - target[i]), 10 - abs(target[i] - _point[i]))
#           visited.add(_point)
#           heapq.heappush(queue, (cost, _point, step))

#     return -1

File: 0752-open-the-lock/0752-open-the-lock.js
// const openLock = (deadends, target) => {
//   const visited = new Set();
//   for (const deadend of deadends) visited.add(deadend);
//   if (visited.has('0000')) return -1;
//   const queue = [['0000', 0]];
//   while (queue.length) {
//     [point, step] = queue.shift()
//     if (point === target) return step;
//     for (let i = 0; i < 4; i++) {
//       for (let s = -1; s < 2; s += 2) {
//         const nb = point.slice(0, i) + Math.abs(10 + ~~point[i] + s) % 10 + point.slice(i + 1);
//         if (!visited.has(nb)) {
//           visited.add(nb)
//           queue.push([nb, step + 1]);
//         }
//       }
//     }
//   }
//   return -1;
// };

class Node {
  constructor(val, prev, next) {
    this.val = val
    this.prev = prev
    this.next = next
  }
}
class Deque {
  constructor() {
    this.tail = null
    this.head = null
    this.size = 0
  }
  append(val) {
    const node = new Node(val, this.tail, null);
    this.tail ? this.tail.next = node : this.head = node;
    this.tail = node;
    this.size++;
  }
  popleft() {
    if (this.size) {
      const node = this.head;
      this.head = this.head.next;
      this.head ? this.head.prev = null : this.tail = null;
      this.size--;
      return node.val;
    }
    return null
  }
}
const openLock = (deadends, target) => {
  const visited = new Set();
  for (const deadend of deadends) visited.add(deadend);
  if (visited.has('0000')) return -1;
  const queue = new Deque();
  queue.append(['0000', 0]);
  while (queue.size) {
    [point, step] = queue.popleft()
    if (point === target) return step;
    for (let i = 0; i < 4; i++) {
      for (let s = -1; s < 2; s += 2) {
        const nb = point.slice(0, i) + Math.abs(10 + ~~point[i] + s) % 10 + point.slice(i + 1);
        if (!visited.has(nb)) {
          visited.add(nb)
          queue.append([nb, step + 1]);
        }
      }
    }
  }
  return -1;
};